src/metabase/domain_entities/converters.cljs

```diff
@@ -1,73 +1,95 @@
+                :tuple             {:enter vec}
+                :map               {:enter decode-map}
+                :map-of            {:compile (fn [schema _]
+                                               (let [[key-s val-s] (mc/children schema)
+                                                     keydec (mc/decoder key-s js-transformer)
+                                                     valdec (mc/decoder val-s js-transformer)]
+                                                 #(decode-map-of keydec valdec %)))}
+                }
+     :encoders {:keyword           name
+                'keyword?          name
+                :qualified-keyword #(str (namespace %) "/" (name %))
+                :vector            {:leave clj->js}
+                :sequential        {:leave clj->js}
+                :tuple             {:leave clj->js}
+                :map               {:compile
+                                    (fn [schema _]
+                                      (let [js-props (into {} (for [[k props] (mc/children schema)
+                                                                    :when (:js/prop props)]
+                                                                [k (:js/prop props)]))
+                                            keyenc   (fn [k] (or (get js-props k)
+                                                                 (csk/->snake_case_string k)))]
+                                        {:leave #(encode-map % keyenc)}))}
+                :map-of            {:leave #(encode-map % name)}}}))
 
-(defmethod incoming :map-of [schema]
-  (let [vf          (incoming (:value schema))
-        keywordize? (-> schema :key :type (#{:keyword 'keyword?}))
-        recursive?  (-> schema :value :type conversion-needed?)
-        transform   (when (and (not (opaque? (:value schema)))
-                               (not= identity vf))
-                      vf)
-        opts        (cond-> {:recursive       (boolean recursive?)
-                             :keywordize-keys (boolean keywordize?)}
-                      keywordize? (assoc :prop->key csk/->kebab-case-keyword
-                                         :key->prop csk/->camelCaseString)
-                      transform   (assoc :transform transform))]
-    #(bean/bean % opts)))
+(defn incoming
+  "Returns a function for converting a JS value into CLJS data structures, based on a schema."
+  [schema]
+  (mc/decoder schema js-transformer))
 
-(defmethod incoming :map [_schema]
-  ->entity)
-
-;; TODO Does this handle cases where we didn't convert string keys to keywords? Does it need to be schema-smart too?
-#_{:clj-kondo/ignore [:clojure-lsp/unused-public-var]} ;; This is used by macro-generated code.
 (defn outgoing
-  "Converts back to pure JS form."
-  [x]
-  (entity-> x))
+  "Returns a function for converting a CLJS value back into a plain JS one, based on its schema."
+  [schema]
+  (mc/encoder schema js-transformer))
```

Comment:
[mc/coercer](https://github.com/metosin/malli#coercion) may be better here too.

Reply:
That way we would get faster failures / a nice assurance that the shapes are what we expect